### PR TITLE
plumb sparse rendering functions

### PR DESCRIPTION
**Pull request by fwilliams**
plumbs sparse rendering functions from C++ to Python and adds unit tests for forward and backward

### CODE REVIEW

File: fvdb/_fvdb_cpp.pyi [hunk 1]
```diff
@@ -205,6 +236,22 @@ class GaussianSplat3d:
         eps_2d: float = ...,
         antialias: bool = ...,
     ) -> tuple[torch.Tensor, torch.Tensor]: ...
+    def render_features_and_depths_sparse(
```
Comment (swahtz) [inline fvdb/_fvdb_cpp.pyi]: Why `render_x_sparse` and not `sparse_render_x`?  I thought we agreed on verb (sparse_render) then noun?  We have `sparse_conv_3d` and `sparse_transpose_conv_3d`, not `conv_3d_sparse`.

File: fvdb/__init__.pyi [hunk 2]
```diff
@@ -103,6 +103,8 @@ __all__ = [
     "GaussianSplat3d",
     "ProjectedGaussianSplats",
     "ConvolutionPlan",
+    "ProjectionType",
```
Comment (swahtz) [inline fvdb/__init__.pyi:104]: Good catch

File: src/fvdb/detail/autograd/GaussianRasterizeSparse.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+#include <fvdb/detail/autograd/GaussianRasterizeSparse.h>
+#include <fvdb/detail/ops/gsplat/GaussianRasterizeBackward.h>
+#include <fvdb/detail/ops/gsplat/GaussianRasterizeForward.h>
+#include <fvdb/detail/utils/Nvtx.h>
+#include <fvdb/detail/utils/Utils.h>
+
+namespace fvdb::detail::autograd {
+
+RasterizeGaussiansToPixelsSparse::VariableList
+RasterizeGaussiansToPixelsSparse::forward(
```
Comment (swahtz) [inline src/fvdb/detail/autograd/GaussianRasterizeSparse.cpp:13]: **💡 suggestion:** ‏I believe this should be 'SparseRasterize' as the verb like 'SparseConv'

Comment (fwilliams) [reply src/fvdb/detail/autograd/GaussianRasterizeSparse.cpp:13]: If we're being ultra pedantic we should say convolve_..._sparse but "SparseConv" is its own concept that is well established in the literature (and used as a noun). I think this is fine.

File: fvdb/_fvdb_cpp.pyi (hunk 1 above)
Comment (fwilliams) [reply fvdb/_fvdb_cpp.pyi]: sparse is an adjective, render is a verb so this conforms. 

The reason to remove is so that when you tab complete, methods that do the same thing but differ in sparsity show up beside each other.

File: src/fvdb/detail/autograd/GaussianRasterizeSparse.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+                                                                       activeTiles,
+                                                                       tilePixelMask,
+                                                                       tilePixelCumsum,
+                                                                       pixelMap,
+                                                                       absgrad);
+    });
+    Variable dLossDMean2dAbs;
+    if (absgrad) {
+        dLossDMean2dAbs = std::get<0>(variables);
+        // means2d.absgrad = dLossDMean2dAbs;
```
Comment (swahtz) [inline src/fvdb/detail/autograd/GaussianRasterizeSparse.cpp]: Can remove

File: tests/unit/test_gaussian_splat_3d.py [hunk 5]
```diff
@@ -2247,11 +2247,367 @@ def test_gaussian_contributors_scene_dense_pixels_sparse_render(self):
...
+
+        dense_features, dense_alphas = self.gs3d.render_features(
+            self.cam_to_world_mats[0:1],
+            self.projection_mats[0:1],
+            self.width,
+            self.height,
+            self.near_plane,  # near_plane
+            self.far_plane,  # far_plane
+        )
+        torch.cuda.synchronize()
```
Comment (swahtz) [inline tests/unit/test_gaussian_splat_3d.py]: **❓ question:** ‏Why the synchronizes here?  Were there any issues without them?

File: tests/unit/test_gaussian_splat_3d.py [hunk 6]
```diff
@@ -2247,11 +2247,367 @@ def test_gaussian_contributors_scene_dense_pixels_sparse_render(self):
...
+            "Sparse quats grad does not match dense quats grad at specified pixels",
+        )
+        self.assertTrue(
+            torch.allclose(sparse_log_scales_grad, dense_log_scales_grad, atol=1e-4, rtol=1e-8),
+            "Sparse log scales grad does not match dense log scales grad at specified pixels",
+        )
+        self.assertTrue(
+            torch.allclose(sparse_logit_opacities_grad, dense_logit_opacities_grad, atol=1e-4, rtol=1e-8),
+            "Sparse logit opacities grad does not match dense logit opacities grad at specified pixels",
+        )
```
Comment (swahtz) [inline tests/unit/test_gaussian_splat_3d.py:2489]: **🚩 issue:** ‏Shouldn't we also be getting and testing the SH grads?

File: tests/unit/test_gaussian_splat_3d.py [hunk 7]
```diff
@@ -2247,11 +2247,367 @@ def test_gaussian_contributors_scene_dense_pixels_sparse_render(self):
...
```
Comment (swahtz) [inline tests/unit/test_gaussian_splat_3d.py:2617]: **❓ question:** ‏As above, shouldn't we be getting and testing the SH grads here?

File: fvdb/_fvdb_cpp.pyi (hunk 1 above)
Comment (swahtz) [reply fvdb/_fvdb_cpp.pyi]: 'Sparse' relates to the rendering, so this would have to be `render_features_and_depths_sparsely` to be grammatically correct and be what we mean.

We also use 'sparse render' throughout other parts of the API (in the functions called further down by this function).

Comment (fwilliams) [reply fvdb/_fvdb_cpp.pyi]: I renamed the other functions in this file too. It's all consistent. Unless there is a good reason, I'd like to keep this naming.

File: tests/unit/test_gaussian_splat_3d.py (hunk 5 above)
Comment (fwilliams) [reply tests/unit/test_gaussian_splat_3d.py]: copilot. I'll rm

File: tests/unit/test_gaussian_splat_3d.py (hunk 6 above)
Comment (fwilliams) [reply tests/unit/test_gaussian_splat_3d.py:2489]: The particular superclass for this test doesn't have requires_grad on these for some reason. I'll add this.

File: src/fvdb/detail/autograd/GaussianRasterizeSparse.cpp (hunk 3 above)
Comment (swahtz) [reply src/fvdb/detail/autograd/GaussianRasterizeSparse.cpp:13]: 'sparse render' is also an established term

File: fvdb/_fvdb_cpp.pyi (hunk 1 above)
Comment (swahtz) [reply fvdb/_fvdb_cpp.pyi]: What's the good reason for changing it?

Comment (swahtz) [reply fvdb/_fvdb_cpp.pyi]: The underlying functions are called things like `GaussianSplat3d::sparseRenderImpl`

Comment (swahtz) [reply fvdb/_fvdb_cpp.pyi]: Why don't we do this, why don't we refactor these functions to eliminate the sparse variants and just make the sparse pixels an optional argument?  So `render_features` just takes a `pixels_to_render` optional arg.  The return types would have to `JaggedTensorOrTensorT` but the return types of the sparse functions is already that.  From what I read, all the other args are the same between the two functions and have the same meanings.

Comment (harrism) [reply fvdb/_fvdb_cpp.pyi]: I will need some time if you want me to weigh in on this debate.

**Review by swahtz — CHANGES_REQUESTED** [review]
Thanks for plumbing this functionality up through the python API and putting in the pytests.  We should @harrism some time to review on his workweek to be able to chat about the naming of the functions or potentially merging the public interface to sparse/dense render into the one function.

File: fvdb/_fvdb_cpp.pyi (hunk 1 above)
Comment (fwilliams) [reply fvdb/_fvdb_cpp.pyi]: For the sake of getting this functionality in, I reverted all naming changes and will make that a seperate PR where we can discuss.

File: fvdb/gaussian_splatting.py [hunk 8]
```diff
@@ -256,12 +257,12 @@ class GaussianSplat3d:
...
       representation (:meth:`render_images`, :meth:`render_images_and_depths`).
     - Rendering depth maps (:meth:`render_depths`, :meth:`render_images_and_depths`).
-    - Rendering features at arbitrary sparse pixel locations (:meth:`sparse_render_features`).
+    - Rendering features at arbitrary sparse pixel locations (:meth:`sparse_render_images`, :meth:`sparse_render_images_and_depths`).
     - Rendering depths at arbitrary sparse pixel locations (:meth:`sparse_render_depths`).
     - Computing which gaussians contribute to each pixel in an image plane
       (:meth:`render_num_contributing_gaussians`, :meth:`render_contributing_gaussian_ids`).
     - Computing the set of Gaussians which contribute to a set of sparse pixel locations
-      (:meth:`sparse_render_num_contributing_gaussians`, :meth:`sparse_render_contributing_gaussian_ids`).
+      (:meth:`render_num_contributing_gaussians_sparse`, :meth:`render_contributing_gaussian_ids_sparse`).
```
Comment (swahtz) [inline fvdb/gaussian_splatting.py]: **⌨️ typo:** ‏I think this revert to the name change was just left should be `sparse_render_contributing...` yes?

File: fvdb/gaussian_splatting.py [hunk 9]
```diff
@@ -1952,7 +2240,7 @@ def render_num_contributing_gaussians(
         )
 
     @overload
-    def sparse_render_num_contributing_gaussians(
+    def render_num_contributing_gaussians_sparse(
```
Comment (swahtz) [inline fvdb/gaussian_splatting.py]: **⌨️ typo:** ‏I think the changes here to this function name (and the ones below) were just left out of the revert, should be `sparse_render_num_contributing_gaussians`, etc.

File: tests/unit/test_gaussian_splat_3d.py [hunk 10]
```diff
@@ -1648,7 +1647,7 @@ def test_gaussians_center_render(self):
         pixels_to_render = JaggedTensor([torch.tensor([[h // 2 - 1, w // 2 - 1]])]).to(self.device)
 
         # test sparse render num contributing gaussians
-        sparse_num_contributing_gaussians, sparse_alphas = gs3d.sparse_render_num_contributing_gaussians(
+        sparse_num_contributing_gaussians, sparse_alphas = gs3d.render_num_contributing_gaussians_sparse(
```
Comment (swahtz) [inline tests/unit/test_gaussian_splat_3d.py]: **⌨️ typo:** ‏Corresponding to the mentions above, I think these just were left out of the naming change revert

File: src/fvdb/detail/autograd/GaussianRasterizeSparse.cpp (hunk 3 above)
Comment (harrism) [reply src/fvdb/detail/autograd/GaussianRasterizeSparse.cpp:13]: In the python code in this PR we have `sparse` first, and in the C++ it is last?

The advantage of putting it last is that like functions and classes get grouped when sorted lexicographically. Of course, that might be a disadvantage to?

File: src/fvdb/GaussianSplat3d.cpp [hunk 11]
```diff
@@ -443,6 +445,47 @@ GaussianSplat3d::renderImpl(const torch::Tensor &worldToCameraMatrices,
         state, settings.tileSize, settings.imageWidth, settings.imageHeight, 0, 0);
 }
 
+std::tuple<JaggedTensor, JaggedTensor>
+GaussianSplat3d::sparseRenderImpl(const JaggedTensor &pixelsToRender,
```
Comment (harrism) [inline src/fvdb/GaussianSplat3d.cpp:449]: Here we have "sparse" first and in GaussianRasterizeSparse.h it is last. In general naming is not consistent in this PR.
